lighty-modules/lighty-gnmi/lighty-gnmi-device-simulator/src/main/java/io/lighty/modules/gnmi/simulatordevice/gnmi/GnmiCrudService.java

```diff
@@ -347,11 +358,67 @@ private Optional<DataSchemaNode> findAugmentationFromOuterModel(final String ele
+    private YangInstanceIdentifier getYIIDWithNewPredicateNode(final YangInstanceIdentifier resultIdentifier,
             final PathElem currentElement) {
         final var qname = resultIdentifier.getLastPathArgument().getNodeType();
         final Map<QName, Object> keysMap = currentElement.getKeyMap().entrySet().stream()
-                .collect(Collectors.toMap(e -> QName.create(qname, e.getKey()), Map.Entry::getValue));
+            .collect(Collectors.toMap(e -> QName.create(qname, e.getKey()), Map.Entry::getValue));
+
+        for (final var entry : keysMap.entrySet()) {
+            final var keyNode = resultIdentifier.node(NodeIdentifierWithPredicates.of(qname, keysMap))
+                .node(entry.getKey());
+            final var baseTypeDef = getBaseTypeDef(keyNode);
+            entry.setValue(mapToCorrectDataType(baseTypeDef, (String) entry.getValue()));
+        }
         return resultIdentifier.node(NodeIdentifierWithPredicates.of(qname, keysMap));
     }
+
+    private TypeDefinition<?> getBaseTypeDef(final YangInstanceIdentifier identifier) {
+        final var nodeAndStack = DataSchemaContextTree.from(context).enterPath(identifier).get();
+        final var dataSchemaNode = nodeAndStack.node().dataSchemaNode();
+        var resultDataSchemaType = ((EmptyLeafEffectiveStatement) dataSchemaNode).getType();
+        if (resultDataSchemaType instanceof LeafrefTypeDefinition leafRefType) {
+            final var leafRefPathOrig = leafRefType.getPathStatement().getOriginalString();
+            final var leafRefPathList = Arrays.stream(leafRefPathOrig.split("/")).filter(s -> !s.isEmpty()).toList();
+            final var stack = nodeAndStack.stack();
+            for (final var path : leafRefPathList) {
+                if (path.equals("..")) {
+                    stack.exit();
+                } else {
+                    stack.enterSchemaTree(QName.create(stack.currentModule().localQNameModule(), path));
+                }
+            }
+            resultDataSchemaType = ((EmptyLeafEffectiveStatement) stack.currentStatement()).getType();
+        }
+        return resultDataSchemaType.getBaseType() != null ? resultDataSchemaType.getBaseType() : resultDataSchemaType;
+    }
+
+    private Object mapToCorrectDataType(final TypeDefinition<?> typeDefinition, final String value) {
+        final var qname = typeDefinition.getQName();
+        if (typeDefinition instanceof IdentityrefTypeDefinition identityType) {
+            final var firstIdentity = identityType.getIdentities().iterator().next();
+            final var identityQname = firstIdentity.getQName();
+            final var splitValue = value.split(":");
+            final var identityRefName = splitValue[splitValue.length - 1];
+            return QName.create(identityQname, identityRefName);
+        } else if (qname.equals(TypeDefinitions.BOOLEAN)) {
+            return value.equals("true");
```

Comment:
I would prefer `Boolean.valueOf(value);`.

Reply:
Done

Reply:
IMO how user can specify "false" when we have `return Boolean.valueOf("true");`?

Reply:
Done